pkg/verification/service/verification_service_test.go

```diff
@@ -896,41 +896,279 @@ func (s *TestVerificationServiceSuite) TestVerifyPhoneCode() {
+						toolchainv1alpha1.UserSignupActivationCounterAnnotationKey: "1", // first time user
+						toolchainv1alpha1.UserVerificationAttemptsAnnotationKey:    "0",
+						toolchainv1alpha1.UserSignupCaptchaScoreAnnotationKey:      "0.5", // and captcha score is low
+						toolchainv1alpha1.UserSignupVerificationCodeAnnotationKey:  "123456",
+						toolchainv1alpha1.UserVerificationExpiryAnnotationKey:      now.Add(10 * time.Second).Format(verificationservice.TimestampLayout),
+					},
+					Labels: map[string]string{
+						toolchainv1alpha1.UserSignupUserPhoneHashLabelKey: "+1NUMBER",
+					},
+				},
+				Spec: toolchainv1alpha1.UserSignupSpec{
+					Username: "sbryzak@redhat.com",
+				},
+			}
+			states.SetVerificationRequired(userSignup, true)
+
+			err := s.FakeUserSignupClient.Delete(userSignup.Name, nil)
+			require.NoError(s.T(), err)
+
+			err = s.FakeUserSignupClient.Tracker.Add(userSignup)
+			require.NoError(s.T(), err)
+
+			ctx, _ := gin.CreateTestContext(httptest.NewRecorder())
+			err = s.Application.VerificationService().VerifyPhoneCode(ctx, userSignup.Name, userSignup.Spec.Username, "123456")
+			require.EqualError(s.T(), err, "verification failed: verification is not available at this time")
+
+			_, err = s.FakeUserSignupClient.Get(userSignup.Name)
+			require.NoError(s.T(), err)
+		})
+
+		t.Run("captcha score is ok", func(t *testing.T) {
+			userSignup := &toolchainv1alpha1.UserSignup{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "123",
+					Namespace: configuration.Namespace(),
+					Annotations: map[string]string{
+						toolchainv1alpha1.UserSignupUserEmailAnnotationKey:         "sbryzak@redhat.com",
+						toolchainv1alpha1.UserSignupActivationCounterAnnotationKey: "1", // first time user
+						toolchainv1alpha1.UserVerificationAttemptsAnnotationKey:    "0",
+						toolchainv1alpha1.UserSignupCaptchaScoreAnnotationKey:      "0.6", // and captcha score is ok
+						toolchainv1alpha1.UserSignupVerificationCodeAnnotationKey:  "123456",
+						toolchainv1alpha1.UserVerificationExpiryAnnotationKey:      now.Add(10 * time.Second).Format(verificationservice.TimestampLayout),
+					},
+					Labels: map[string]string{
+						toolchainv1alpha1.UserSignupUserPhoneHashLabelKey: "+1NUMBER",
+					},
+				},
+				Spec: toolchainv1alpha1.UserSignupSpec{
+					Username: "sbryzak@redhat.com",
+				},
+			}
+			states.SetVerificationRequired(userSignup, true)
+
+			err := s.FakeUserSignupClient.Delete(userSignup.Name, nil)
+			require.NoError(s.T(), err)
+
+			err = s.FakeUserSignupClient.Tracker.Add(userSignup)
+			require.NoError(s.T(), err)
+
+			ctx, _ := gin.CreateTestContext(httptest.NewRecorder())
+			err = s.Application.VerificationService().VerifyPhoneCode(ctx, userSignup.Name, userSignup.Spec.Username, "123456")
+			require.NoError(s.T(), err)
+
+			userSignup, err = s.FakeUserSignupClient.Get(userSignup.Name)
+			require.NoError(s.T(), err)
+
+			require.False(s.T(), states.VerificationRequired(userSignup))
+		})
```

Comment:
Alternatively you could organize the tests as an array of cases where you provide:
- the score and activation counter annotations
- configuration
- expected result
And then go over them to test. It will reduce code duplication and will be easy to add more cases.

Could you please also add tests to cover cases with:
- there is no score annotation at all.
- score annotation is not valid

Reply:
> Could you please also add tests to cover cases 

Thanks! Done in the same commit. But I'm not 100% sure what should happen in those cases, seems like currently the verification process just goes on, ignoring that the captcha score is missing/invalid. So I just did that. 

Not sure if we should log/return an error or if this is already being done in other parts of the verification process, I guess before going to the phone verification process.

Reply:
If the score is missing that we should treat it as a high score. For invalid scores ideally we would want to return an error but it's OK if we just log the error and ignore it for now.